src/main/kotlin/parts/wisdom/arcraider/visualization/ARCWindow.kt

```diff
@@ -1,56 +1,143 @@
+                        add(createGridBox("Output", outputGrids))
+                        add(Box.createRigidArea(Dimension(GAP_SIZE, 0)))
+                    }
+            )
+            add(Box.createVerticalGlue())
         }
-        val buttonPanel = renderButtonPanel(content)
+    }
 
-        content.add(buttonPanel)
-        content.add(gridPane)
-        add(content)
+    private fun refresh(task: Task) {
+        content.removeAll()
+        renderTask(task)
+        content.revalidate()
+        content.repaint()
     }
 
-    private fun renderButtonPanel(content : JPanel) : JPanel {
-        val buttonPanel = JPanel(FlowLayout(FlowLayout.LEFT)).apply {
-            preferredSize = Dimension(800, 60)
-        }
+    private fun createGridBox(label: String, grids: List<GridPane>) =
+            Box.createVerticalBox().apply {
+                add(
+                        Box.createHorizontalBox().apply {
+                            add(Box.createHorizontalGlue())
+                            add(JLabel(label))
+                            add(Box.createHorizontalGlue())
+                        }
+                )
+                add(Box.createRigidArea(Dimension(0, GAP_SIZE / 2)))
+                grids.forEach {
+                    add(it)
+                    add(Box.createRigidArea(Dimension(0, GAP_SIZE / 2)))
+                }
+            }
 
-        val chooserButton = JButton("Browse").apply {
-            preferredSize = Dimension(100, 40)
+    private fun createBrowseButton(): JButton {
+        return JButton("Browse").apply {
             addActionListener {
                 val fileChooser = JFileChooser()
-                val result = fileChooser.showOpenDialog(content)
+                val result = fileChooser.showOpenDialog(null)
                 if (result == JFileChooser.APPROVE_OPTION) {
                     val task = loadTaskFromFile(fileChooser.selectedFile)
-                    gridPane.resetGrid(Grid(task.test[0].input))
+                    refresh(task)
                 }
             }
         }
+    }
 
-        buttonPanel.add(chooserButton)
-        return buttonPanel
+    private fun createToggle(task: Task, taskType: TaskType): JToggleButton {
+        return JToggleButton(taskType.label, this.taskType == taskType).apply {
+            addActionListener {
+                if (this@ARCWindow.taskType != taskType) {
+                    this@ARCWindow.taskType = taskType
+                    refresh(task)
+                } else {
+                    // Prevent toggling if we are already in this state.
+                    isSelected = true
+                }
+            }
+        }
+    }
+
+    private fun createFromTask(task: Task): TaskGrids {
+        fun gridPane(serializedGrid: SerializedGrid): GridPane {
+            return GridPane(Grid(serializedGrid)).apply {
+                preferredSize = Dimension(GRID_SIZE, GRID_SIZE)
```

Comment:
`GridPane`s  should not be as "stretchy" as these end up being. I'm not quite sure what the rules should be, but see for example [this PR's rendering](https://snipboard.io/uRFEJb.jpg) of task e98196ab versus the official [ARC testing interface](https://snipboard.io/D7c340.jpg).

I think it's fine to improve this either in this PR or in a subsequent PR.

Reply:
I'll address this in #13